File: java/android/benchmarks/src/androidTest/java/GroupSendEndorsements.java
//
// Copyright 2024 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import androidx.benchmark.BenchmarkState;
import androidx.benchmark.junit4.BenchmarkRule;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;
import org.signal.libsignal.protocol.ServiceId;
import org.signal.libsignal.zkgroup.ServerPublicParams;
import org.signal.libsignal.zkgroup.ServerSecretParams;
import org.signal.libsignal.zkgroup.VerificationFailedException;
import org.signal.libsignal.zkgroup.groups.ClientZkGroupCipher;
import org.signal.libsignal.zkgroup.groups.GroupSecretParams;
import org.signal.libsignal.zkgroup.groups.UuidCiphertext;
import org.signal.libsignal.zkgroup.groupsend.GroupSendDerivedKeyPair;
import org.signal.libsignal.zkgroup.groupsend.GroupSendEndorsement;
import org.signal.libsignal.zkgroup.groupsend.GroupSendEndorsementsResponse;

@RunWith(Parameterized.class)
public class GroupSendEndorsements {
  @Rule public final BenchmarkRule benchmarkRule = new BenchmarkRule();

  @Parameters(name = "groupSize={0}")
  public static Object[] data() {
    return new Integer[] {10, 100, 1000};
  }

  private final ServerSecretParams serverParams = ServerSecretParams.generate();
  private final ServerPublicParams serverPublicParams = serverParams.getPublicParams();
  private final GroupSecretParams groupParams = GroupSecretParams.generate();

  private final Instant expiration =
      Instant.now().truncatedTo(ChronoUnit.DAYS).plus(2, ChronoUnit.DAYS);

  private final ServiceId.Aci[] members;
  private final UuidCiphertext[] encryptedMembers;
  private final GroupSendEndorsementsResponse response;

  public GroupSendEndorsements(int groupSize) {
    members = new ServiceId.Aci[groupSize];
    for (int i = 0; i < groupSize; ++i) {
      members[i] = new ServiceId.Aci(UUID.randomUUID());
    }

    encryptedMembers = new UuidCiphertext[groupSize];
    final ClientZkGroupCipher cipher = new ClientZkGroupCipher(groupParams);
    for (int i = 0; i < groupSize; ++i) {
      encryptedMembers[i] = cipher.encrypt(members[i]);
    }

    GroupSendDerivedKeyPair keyPair =
        GroupSendDerivedKeyPair.forExpiration(expiration, serverParams);
    response = GroupSendEndorsementsResponse.issue(Arrays.asList(encryptedMembers), keyPair);
  }

  @Test
  public void benchmarkReceiveWithServiceIds() throws VerificationFailedException {
    final BenchmarkState state = benchmarkRule.getState();

    while (state.keepRunning()) {
      response.receive(Arrays.asList(members), members[0], groupParams, serverPublicParams);
    }
  }

  @Test
  public void benchmarkReceiveWithCiphertexts() throws VerificationFailedException {
    final BenchmarkState state = benchmarkRule.getState();

    while (state.keepRunning()) {
      response.receive(Arrays.asList(encryptedMembers), encryptedMembers[0], serverPublicParams);
    }
  }

  @Test
  public void benchmarkToToken() throws VerificationFailedException {
    final BenchmarkState state = benchmarkRule.getState();
    final List<GroupSendEndorsement> endorsements =
        response
            .receive(Arrays.asList(encryptedMembers), encryptedMembers[0], serverPublicParams)
            .endorsements();

    while (state.keepRunning()) {
      for (GroupSendEndorsement next : endorsements) {
        next.toToken(groupParams);
      }
    }
  }
}

File: swift/Benchmarks/Sources/Benchmarks/GroupSendEndorsements.swift
//
// Copyright 2024 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Benchmark
import Foundation
import LibSignalClient

private let SECONDS_PER_DAY: UInt64 = 24 * 60 * 60

let groupSendEndorsementsSuite = BenchmarkSuite(name: "GroupSendEndorsements") { suite in
    let serverParams = try! ServerSecretParams.generate()
    let serverPublicParams = try! serverParams.getPublicParams()
    let groupParams = try! GroupSecretParams.generate()
    let now = UInt64(Date().timeIntervalSince1970)
    let startOfDay = now - now % SECONDS_PER_DAY
    let expiration = Date(timeIntervalSince1970: TimeInterval(startOfDay)).addingTimeInterval(TimeInterval(2 * SECONDS_PER_DAY))

    for groupSize in [10, 100, 1000] {
        let members = (0..<groupSize).map { _ in Aci(fromUUID: UUID()) }
        let cipher = ClientZkGroupCipher(groupSecretParams: groupParams)
        let encryptedMembers = members.map { try! cipher.encrypt($0) }

        let keyPair = GroupSendDerivedKeyPair.forExpiration(expiration, params: serverParams)
        let response = GroupSendEndorsementsResponse.issue(groupMembers: encryptedMembers, keyPair: keyPair)

        suite.benchmark("receiveWithServiceIds/\(groupSize)") {
            blackHole(try! response.receive(groupMembers: members, localUser: members[0], groupParams: groupParams, serverParams: serverPublicParams))
        }
        suite.benchmark("receiveWithCiphertexts/\(groupSize)") {
            blackHole(try! response.receive(groupMembers: encryptedMembers, localUser: encryptedMembers[0], serverParams: serverPublicParams))
        }

        let endorsements = try! response.receive(groupMembers: members, localUser: members[0], groupParams: groupParams, serverParams: serverPublicParams)

        suite.benchmark("toToken/\(groupSize)") {
            blackHole(endorsements.endorsements.map { $0.toToken(groupParams: groupParams) })
        }
    }
}
